src/tss2-tcti/tcti-spi.c

```diff
@@ -0,0 +1,612 @@
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <assert.h>
+
+#include "tss2_tcti.h"
+#include "tss2_tcti_spi.h"
+#include "tss2_mu.h"
+#include "tcti-common.h"
+#include "tcti-spi.h"
+#include "util/io.h"
+#define LOGMODULE tcti
+#include "util/log.h"
+
+static void spi_tpm_delay_ms(TSS2_TCTI_SPI_CONTEXT* ctx, int milliseconds)
+{
+    // Sleep a specified amount of milliseconds
+    ctx->platform.sleep_ms(ctx->platform.user_data, milliseconds);
+}
+
+static void spi_tpm_start_timeout(TSS2_TCTI_SPI_CONTEXT* ctx, int milliseconds)
+{
+    // Start a timeout timer with the specified amount of milliseconds
+    ctx->platform.start_timeout(&ctx->platform.user_data, milliseconds);
+}
+
+static bool spi_tpm_timeout_expired(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    // Check if the last started tiemout expired
+    return ctx->platform.timeout_expired(ctx->platform.user_data);
+}
+
+static TSS2_RC spi_tpm_spi_acquire(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    // Reserve SPI bus until transaction is over and keep pulling CS
+    return ctx->platform.spi_acquire(ctx->platform.user_data);
+} 
+
+static TSS2_RC spi_tpm_spi_release(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    // Release SPI bus and release CS
+    return ctx->platform.spi_release(ctx->platform.user_data);
+} 
+
+static TSS2_RC spi_tpm_spi_transfer(TSS2_TCTI_SPI_CONTEXT* ctx, const void *data_out, void *data_in, size_t cnt)
+{
+    // Perform SPI transaction with cnt bytes
+    return ctx->platform.spi_transfer(ctx->platform.user_data, data_out, data_in, cnt);
+}
+
+static void spi_tpm_platform_finalize(TSS2_TCTI_SPI_CONTEXT* ctx)
+{
+    // Free user_data and resources inside
+    ctx->platform.finalize(ctx->platform.user_data);
+}
+
+static inline uint32_t spi_tpm_read_be32(const void *src)
+{
+    const uint8_t *s = src;
+    return (((uint32_t)s[0]) << 24) | (((uint32_t)s[1]) << 16) | (((uint32_t)s[2]) << 8) | (((uint32_t)s[3]) << 0);
+}
+
+static TSS2_RC spi_tpm_start_transaction(TSS2_TCTI_SPI_CONTEXT* ctx, enum TCTI_SPI_REGISTER_ACCESS_TYPE access, size_t bytes, uint32_t addr)
+{
+    TSS2_RC rc;
+
+    // Build spi header
+    uint8_t header[4];
+    
+    // Transaction type and transfer size
+    header[0] = ((access == TCTI_SPI_REGISTER_READ) ? 0x80 : 0x00) | (bytes - 1);
+    
+    // TPM register address
+    header[1] = addr >> 16 & 0xff;
+    header[2] = addr >> 8  & 0xff;
+    header[3] = addr >> 0  & 0xff;
+
+    // Reserve SPI bus until transaction is over and keep pulling CS
+    rc = spi_tpm_spi_acquire(ctx);
+    if (rc != TSS2_RC_SUCCESS) {
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    // Send header
+    uint8_t header_response[4];
+    rc = spi_tpm_spi_transfer(ctx, header, header_response, 4);
+    if (rc != TSS2_RC_SUCCESS) {
+        return TSS2_TCTI_RC_IO_ERROR;
+    }
+
+    // Wait until the TPM exits the wait state and sends a 1 bit
```

Comment:
can be consolidated into one chunk, by movinf the byte&1 to the begining of the loop.

Reply:
Yes, but then we would not check the last read byte, right?!
For me this "fast path" before the retry loop also conveys a better intuition of the protocol when reading the code.

Reply:
I don't get the first part about.
I'm fine with the  current implementation.

Reply:
Okay, I lowered the retry count to 256 and added a comment.